Image: Adds the block controls for uploading image.

vipul0425 opened the pull request:
## What?
Fixes #54867

## Why?
In small width size the image block placeholder looks broken and it's hard to choose an image from there.

## How?
This PR is adding a BlockControl to choose an image which will have all the three options upload, choose from media library or add from a link and will keep the placeholder abstract.

## Testing Instructions
1. Add a columns block or grid block. Choose 6+ columns or grid in a row and add an image block.
2. Now the choose an image option will appear in BlockControls Toolbar.  

## Screenshots or screencast

For Large Size
<img width="676" alt="image" src="https://github.com/user-attachments/assets/2161cd0c-5a1e-4f2d-8511-a8e928825661">


For Medium Size
<img width="534" alt="image" src="https://github.com/user-attachments/assets/a339a420-8e20-4a15-9335-4844df6beb04">

For small Size
<img width="483" alt="image" src="https://github.com/user-attachments/assets/1b88635f-df3b-452f-9fcf-72d6b26f45ec">

vipul0425 committed:
feat: Adds the block controls for uploading image.

vipul0425 committed:
feat: Adds the size based placeholder logic.

jasmussen commented:
Nice! This makes good progress, but doesn't yet fully fix 54867. Quick GIF:

![Image](https://github.com/user-attachments/assets/347111cf-537e-4d9e-8bfe-995985526da4)

We need only two changes:

1. Rename "Choose Image" to "Add image" (note the lowercase i)
2. Don't show the "white material" in mobile.

2 should be possible with CSS. Essentially it should look like this:

![Screenshot 2024-08-13 at 13 48 37](https://github.com/user-attachments/assets/65d335e1-ad43-48a4-8ecd-09325bc79aa4)

That maintains the gray placeholder material rather than the white `<Placeholder />` material on-select. You should be able to use the CSS class already present, for this. Something like:

* Change `.wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder` to `.wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder:not(.is-small)`
* Change `.wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder .components-placeholder__illustration` to .wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder .components-placeholder__illustration:not(.is-small)`
* Change `.wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder:before` to `.wp-block-image.wp-block-image.is-selected .block-editor-media-placeholder:not(.is-small):before`
* Change `.is-selected>.components-placeholder.has-illustration .components-placeholder__label` to `.is-selected>.components-placeholder.has-illustration:not(.is-small) .components-placeholder__label`

There are probably better places to do this, this was some quick inspector work. Let me know if you'd like me to try and push some code.

vipul0425 committed:
fix: The layout design for small images and adds grid support.

jasmussen commented:
Nice. Took the latest for a spin:

![status](https://github.com/user-attachments/assets/2a84c2bc-2dd1-4786-846c-ee4aeb82195c)

In the above I added a `min-height: 60px;` to the placeholder component. The default state is a bit small:

![Screenshot 2024-08-14 at 10 16 41](https://github.com/user-attachments/assets/067f5056-3ffd-44fd-b8dc-bf53b28c2707)

Here's 60px again:

![Screenshot 2024-08-14 at 10 14 22](https://github.com/user-attachments/assets/8ca970fd-e179-47fc-8596-d9841bb5890a)

CC: @WordPress/gutenberg-design in case you have opinions on the min-height of this state. 

I noticed one glitch. when you insert the image in a narrow column, there's a brief flash of the white placeholder material:
![insert](https://github.com/user-attachments/assets/aee02f3f-7d36-430d-ae3b-bd26621a72d0)

It's a very brief flash, but enough that it's annoying. Since the CSS would never show it, I wonder if it's the resize-observe applying the is-small CSS class with a slight delay? Is there any trick we can do here?

Separately, I noticed that this PR adds a resizeobserver locally to the Image block. That may be fine (perhaps @WordPress/gutenberg-components would know whether this is best done directly in the Placeholder component or if it's good to keep here). I'm asking mainly because the `Placeholder` component already has one, and it already applies a CSS class. Could we rely on that existing CSS class? It's fine that _every placeholder_ gets the is-small changes you've made here, even if only the Image block makes use of them for now. Make sense?

jasmussen commented on packages/block-library/src/image/edit.js:
```diff
@@ -107,8 +111,16 @@ export function ImageEdit( {
+
 	const [ temporaryURL, setTemporaryURL ] = useState( attributes.blob );
 
+	const [ resizeListener, sizes ] = useResizeObserver();
```
Pretty sure this one is already included in the `Placeholder` component, so it may be best to make these width changes to that component itself, rather than locally here.

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -364,35 +380,72 @@ export function ImageEdit( {
+				{ resizeListener }
+				{ ! ( temporaryURL || url ) && (
+					<BlockControls group="other">
+						<MediaReplaceFlow
```
Is it possible to integrate this code into [the existing `MediaReplaceFlow` component](https://github.com/WordPress/gutenberg/blob/00d1f8afac2e3b9764dc28261462b72794ae5ce0/packages/block-library/src/image/image.js#L590-L602)?

The reason is that, for example, if multiple blocks are selected or if the URL control is locked, the `MediaReplaceFlow` component should not be rendered.

For example, try inserting the following HTML via the code editor:

```html
<!-- wp:image { "metadata":{ "bindings":{ "url":{ "source":"core/post-meta", "args":{ "key":"test" } } } } } -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image -->
```


In this case, the "Add image" control should not be displayed:

![image](https://github.com/user-attachments/assets/d19c0445-9a7d-4bdf-9018-a5b929960cf9)

The existing `MediaReplaceFlow` component has checks for these:

https://github.com/WordPress/gutenberg/blob/00d1f8afac2e3b9764dc28261462b72794ae5ce0/packages/block-library/src/image/image.js#L590

Furthermore, in this PR, it seems that it is no longer possible to replace images by dropping them, as shown by the E2E test failure. This might be related to the fact that the `MediaReplaceFlow` component is separated:

https://github.com/user-attachments/assets/0df35c35-981e-4825-9c62-230d456d7e0c

mirka commented on packages/block-library/src/image/edit.js:
```diff
@@ -107,8 +111,16 @@ export function ImageEdit( {
+
 	const [ temporaryURL, setTemporaryURL ] = useState( attributes.blob );
 
+	const [ resizeListener, sizes ] = useResizeObserver();
```
> Could we rely on that existing CSS class? It's fine that _every placeholder_ gets the is-small changes you've made here, even if only the Image block makes use of them for now. Make sense?

Technically, yes. But in the long run, I'm thinking it could be cleaner and more flexible if this kind of responsive styling was handled solely by the consumer of Placeholder, using [CSS container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) (rather than a resize observer). That way, consumers wouldn't have to depend on a set of arbitrary breakpoints that Placeholder defines.

I'm not intimately familiar with the requirements of this Image block in particular, but on first glance I would say it could make it simpler and more performant. Could be worth trying here. (This is a non-blocking comment, just suggesting in case it hadn't been considered.)

vipul0425 replied:
Hi @mirka, this would be a good change, but since it applies more generally to the placeholder component, should we make the change in this PR or address it in a separate issue? Related: https://github.com/WordPress/gutenberg/issues/64288

ciampo replied:
Agreed with Lena — this seems a good fit for CSS Container Queries

_(edit: when I wrote my message, I hadn't seen @vipul0425 's reply, which was posted almost at the same time)_

mirka replied:
@vipul0425 I don't have much insight into the current state of things, but it looks like you were able to achieve what you needed in this PR without resize observers or container queries, so yes it seems like something that can be addressed as a separate issue 👍

vipul0425 committed:
refactor: Remove resize observer and add lockURLControls support.

vipul0425 commented:
Hi @jasmussen, I have removed the ResizeObserver from this component and refactored the logic using CSS. I also added a min-height for the small variant.

@t-hamano, I have implemented the lockUrlControls logic in the external MediaReplaceFlow, and now we can also replace images. However, my only concern is that in the small size, the "Connected to post meta" text (lockUrlControlsMessage) is also hidden. Do you have any suggestions on where we can display this? I think there should be some visual distinction apart from just the purple image icon that is visible in the toolbar for this.

<img width="342" alt="image" src="https://github.com/user-attachments/assets/e6eacf77-cb92-42c9-8183-34253e894883">

<img width="391" alt="image" src="https://github.com/user-attachments/assets/54feb64f-dd7f-4ac3-b01a-1004eba8ca0a">

jasmussen commented:
> However, my only concern is that in the small size, the "Connected to post meta" text (lockUrlControlsMessage) is also hidden. Do you have any suggestions on where we can display this? I think there should be some visual distinction apart from just the purple image icon that is visible in the toolbar for this.

We should show this in the inspector, if it isn't already. If it's already shown there, it's okay to not show it here, and just show the same gray placeholder material.

Taking a look at the branch again now, thanks for your continued work!

jasmussen commented:
Works well!

![Screenshot 2024-08-15 at 09 57 42](https://github.com/user-attachments/assets/0b52f896-40e5-45aa-abe8-5c8ffab35f31)

I'll defer to others for a code review. I also imagine that the brief flash of the white material is hard to fully get rid of. I wonder if—and this can be a separate followup PR—the logic should be _inversed_: instead of _hiding_ the elaborate white placeholder states in narrow contexts, we _show_ them in the wider contexts? Perhaps the flash would then also be inverted, so this might not be the best solution. But something to potentially try!

ciampo commented on packages/block-library/src/image/editor.scss:
```diff
@@ -27,13 +26,25 @@
+		min-height: 60px;
+	}
+
+	.block-editor-media-placeholder:not(.is-small) {
+		.components-placeholder__fieldset,
+		.components-placeholder__label {
+			display: flex;
+		}
```
`.components-*` class names are meant to be a private implementation detail and should not be used outside of the components package.

Overrides like the one highlighter make this usage `Placeholder` component prone to future breakage, should the component update internally (either its classnames, or how its internal layout is organized).

Having said that, I can see that there are more examples of using private classnames and in general applying style overrides in this file, which is a strong indicator that we should completely rethink how the media placeholder is implemented. So, in that sense, my feedback shouldn't considered blocking for this specific PR, but rather a warning about how badly styles are architected for this component and the risks.

It feels like it's trying to bend the `Placeholder` component beyond its limits. Therefore:

- we should determine if there are any features that we can add to the `Placeholder` component so that consumers don't have to resort to style overrides and other hacky practices;
- otherwise, we should re-implement the component in a way that builds _on top_ of `Placeholder` . For example, instead of setting `display:none` to `.components-placeholder__illustration`, why don't we pass the `withIllustration={false}` prop?
- finally, if the designs / feature are diverging too much, we should ask ourselves whether this warrants a new, separate component (in the block library package)

t-hamano commented:
@vipul0425 Thanks for the update!

> I have implemented the `lockUrlControls` logic in the external `MediaReplaceFlow`, and now we can also replace images.

Sorry if my suggestion was unclear.

What I meant is, I don't think we need to add another `MediaReplaceFlow` component. I mean, I think we can update [here](https://github.com/WordPress/gutenberg/blob/8fc3df806a7e3494937477155fad18f92fdb9538/packages/block-library/src/image/image.js#L592-L600) and add a prop like `name={ ! url ? __( 'Add image' ) : __( 'Replace' ) }`. Such an implementation is also done [here](https://github.com/WordPress/gutenberg/blob/8fc3df806a7e3494937477155fad18f92fdb9538/packages/block-library/src/cover/edit/block-controls.js#L103).

vipul0425 commented:
@t-hamano I tried to integrate that code [here](https://github.com/WordPress/gutenberg/blob/8fc3df806a7e3494937477155fad18f92fdb9538/packages/block-library/src/image/image.js#L592-L600), but since it is defined within the Image component, which isn't rendered on the first load, the inspector controls aren't visible. Alternatively, we could define MediaReplace at the root level and reuse it in both the cases. Could you please guide me on whether that approach would work?

t-hamano commented:
@vipul0425 

How about making the following changes to this PR?

- Extract the `MediaReplaceFlow` component from the `control` component as `mediaReplaceFlow`
- Make sure `mediaReplaceFlow` is always rendered

<details><summary>Diff</summary>

```diff
diff --git a/packages/block-library/src/image/edit.js b/packages/block-library/src/image/edit.js
index c2b1755670..e867461f00 100644
--- a/packages/block-library/src/image/edit.js
+++ b/packages/block-library/src/image/edit.js
@@ -375,34 +375,6 @@ export function ImageEdit( {
        return (
                <>
                        <figure { ...blockProps }>
-                               { ! ( temporaryURL || url ) && ! lockUrlControls && (
-                                       <BlockControls group="other">
-                                               <MediaReplaceFlow
-                                                       mediaId={ id }
-                                                       mediaURL={ url }
-                                                       allowedTypes={ ALLOWED_MEDIA_TYPES }
-                                                       accept="image/*"
-                                                       name={ __( 'Add image' ) }
-                                                       onSelect={ onSelectImage }
-                                                       onError={ onUploadError }
-                                               >
-                                                       <form className="block-editor-media-flow__url-input has-siblings">
-                                                               <span className="block-editor-media-replace-flow__image-url-label">
-                                                                       { __( 'Insert from URL:' ) }
-                                                               </span>
-
-                                                               <LinkControl
-                                                                       value={ { url } }
-                                                                       settings={ [] }
-                                                                       showSuggestions={ false }
-                                                                       onChange={ ( value ) => {
-                                                                               onSelectURL( value.url );
-                                                                       } }
-                                                               />
-                                                       </form>
-                                               </MediaReplaceFlow>
-                                       </BlockControls>
-                               ) }
                                <Image
                                        temporaryURL={ temporaryURL }
                                        attributes={ attributes }
diff --git a/packages/block-library/src/image/image.js b/packages/block-library/src/image/image.js
index 7dd03a7fb5..775482f0ab 100644
--- a/packages/block-library/src/image/image.js
+++ b/packages/block-library/src/image/image.js
@@ -551,6 +551,27 @@ export default function Image( {
 
        const showBlockControls = showUrlInput || allowCrop || showCoverControls;
 
+       const mediaReplaceFlow = (
+               <>
+                       { console.log(
+                               isSingleSelected && ! isEditingImage && ! lockUrlControls
+                       ) }
+                       { isSingleSelected && ! isEditingImage && ! lockUrlControls && (
+                               <BlockControls group="other">
+                                       <MediaReplaceFlow
+                                               mediaId={ id }
+                                               mediaURL={ url }
+                                               name={ ! url ? __( 'Add image' ) : __( 'Replace' ) }
+                                               allowedTypes={ ALLOWED_MEDIA_TYPES }
+                                               accept="image/*"
+                                               onSelect={ onSelectImage }
+                                               onSelectURL={ onSelectURL }
+                                               onError={ onUploadError }
+                                       />
+                               </BlockControls>
+                       ) }
+               </>
+       );
        const controls = (
                <>
                        { showBlockControls && (
@@ -587,19 +608,6 @@ export default function Image( {
                                        ) }
                                </BlockControls>
                        ) }
-                       { isSingleSelected && ! isEditingImage && ! lockUrlControls && (
-                               <BlockControls group="other">
-                                       <MediaReplaceFlow
-                                               mediaId={ id }
-                                               mediaURL={ url }
-                                               allowedTypes={ ALLOWED_MEDIA_TYPES }
-                                               accept="image/*"
-                                               onSelect={ onSelectImage }
-                                               onSelectURL={ onSelectURL }
-                                               onError={ onUploadError }
-                                       />
-                               </BlockControls>
-                       ) }
                        { isSingleSelected && externalBlob && (
                                <BlockControls>
                                        <ToolbarGroup>
@@ -1001,12 +1009,18 @@ export default function Image( {
        }
 
        if ( ! url && ! temporaryURL ) {
-               // Add all controls if the image attributes are connected.
-               return metadata?.bindings ? controls : sizeControls;
+               return (
+                       <>
+                               { mediaReplaceFlow }
+                               { /* Add all controls if the image attributes are connected. */ }
+                               { metadata?.bindings ? controls : sizeControls }
+                       </>
+               );
        }
 
        return (
                <>
+                       { mediaReplaceFlow }
                        { controls }
                        { img }
```

</details>

vipul0425 committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feat/issue-64288

vipul0425 committed:
refactor: Make use of the existing MediaReplaceFlow component.

vipul0425 committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feat/issue-64288

vipul0425 committed:
fix: The mediaPlaceholder location after merge.

vipul0425 committed:
fix: The initial flash in small size.

vipul0425 commented:
Thanks @t-hamano I have worked on it as per your suggestion. I apologize for the delay, it was a busy week. Could you please review the changes at your convenience?

Also @jasmussen I tried to remove that initial flash by reversing the logic but couldn't completely remove that can you please look into if this will work for now. 

Thanks 🙇

jasmussen reviewed (approved):
Impressive work, I don't see the flash anymore:

![img placeholder](https://github.com/user-attachments/assets/3988459c-4837-4de7-a20a-93ac4d648589)

Regardless, this is a substantial improvement!

Needs a quick code check too, but experience wise, this is good to go.

tyxla reviewed (commented):
The failing e2e tests seem to be related, they're worth taking a look at and fixing before moving forward.

t-hamano commented:
Thanks for the update! This PR is working well, but I think we need to find a way to avoid overriding many component styles.

From my understanding, the expected behavior of the placeholder in this PR is as follows:

- Display the illustration when the Image block is not selected
- When the placeholder is small, i.e. when it has the `is-small` class, the illustration will always be displayed regardless of whether the Image block is selected or not.

Currently, [the `withIllustration` prop is always true](https://github.com/vipul0425/gutenberg/blob/911e938d480c743fb47219120e666b132180ad6e/packages/block-library/src/image/edit.js#L339), so there are quite a few CSS overrides to achieve the above conditions.

Ideally, I think we need to _dynamically_ control the `withIllustration` prop. If we could do that, we should be able to remove all the CSS that exists [here](https://github.com/vipul0425/gutenberg/blob/911e938d480c743fb47219120e666b132180ad6e/packages/block-library/src/image/editor.scss#L3-L71).

That is, I want to achieve something like this:

```javascript
export function ImageEdit( { isSelected: isSingleSelected } ) {
	// ...

	const placeholder = ( content ) => {
		// There is no function called getBlockWidth...
		const containerWidth = getBlockWidth();
		const isSmallContainer = containerWidth < 160;

		return (
			<Placeholder
				withIllustration={ ! isSingleSelected || isSmallContainer }
				label={ __( 'Image' ) }
			>
				{ content }
			</Placeholder>
		);
	};

	// ...
}
```

As far as I know, the only way to know the width of a placeholder from outside the placeholder is to use the `useResizeObserver` hook.

@WordPress/gutenberg-components If you know of any other way to achieve the above approach other than using useResizeObserver, please let us know 🙏

mirka commented:
> As far as I know, the only way to know the width of a placeholder from outside the placeholder is to use the `useResizeObserver` hook.

Sounds good to me for the purpose of this PR, and the amount of custom CSS we can shave off seems worth it. We should maybe consider abstracting it into Placeholder itself somehow, but for now a resize observer here feels warranted.

t-hamano commented:
@mirka Thanks for the feedback!

@vipul0425 Sorry for changing the approach so often, but could you try the approach using `useResizeObserver`?

- Remove most (or all) of [this CSS](https://github.com/vipul0425/gutenberg/blob/911e938d480c743fb47219120e666b132180ad6e/packages/block-library/src/image/editor.scss#L3-L71)
- Use the `useResizeObserver` hook to get the width of the `Placeholder` component
- If the width is below a certain value, consider it a small container size
- Dynamically enable the `withIllustration` prop of the Placeholder component (`withIllustration={ ! isSingleSelected || isSmallContainer }`)
- Some kind of workaround may be necessary to prevent the glitch pointed out in [this comment](https://github.com/WordPress/gutenberg/pull/64320#issuecomment-2288133228).

vipul0425 committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feat/issue-64288

vipul0425 committed:
feat: Removes the redundant css and redefine the placeholder logic.

vipul0425 commented:
Hi @t-hamano, I've updated the logic as you suggested. I used the reverse logic for the small container to prevent the initial flash of content. However, I noticed that this is causing two e2e tests to fail. If I'm understanding this correctly, this is due to the focus shifting to the image component instead of the upload button on initial render.

If I remove the isLargeContainer && content condition inside the placeholder, the focus shifts back to the upload button, but then the buttons also appear on the small container, which we don't want.

Do you have any suggestions? I tried manually focusing on the upload button when the component loads, but that approach didn't seem ideal.

t-hamano commented on packages/block-library/src/image/editor.scss:
```diff
@@ -1,39 +1,13 @@
+		min-height: 60px;
 	}
+
 	.block-bindings-media-placeholder-message {
 		opacity: 0;
 	}
```
We also need to remove this CSS, otherwise the message will not be visible when the source is bind:

![image](https://github.com/user-attachments/assets/42654127-91a4-4da7-86ae-47983a04342a)

t-hamano commented on packages/block-library/src/image/editor.scss:
```diff
@@ -1,39 +1,13 @@
-		opacity: 1;
-	}
 
 	// Remove the transition while we still have a legacy placeholder style.
 	// Otherwise the content jumps between the 1px placeholder border, and any inherited custom
```
We should be able to delete this one (`transition: none`) too. See the discussion in #64861 for the reason.

t-hamano commented on packages/block-library/src/image/editor.scss:
```diff
@@ -100,7 +74,9 @@ figure.wp-block-image:not(.wp-block) {
 }
 
-.block-editor-block-list__block[data-type="core/image"] .block-editor-block-toolbar .block-editor-url-input__button-modal {
+.block-editor-block-list__block[data-type="core/image"]
+.block-editor-block-toolbar
+.block-editor-url-input__button-modal {
```
Personally, I find this line break confusing because it makes it hard to understand at a glance whether these selectors are one or separate.

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -330,40 +358,49 @@ export function ImageEdit( {
+						...shadowProps.style,
+					} }
+				>
+					{ lockUrlControls ? (
+						<span className="block-bindings-media-placeholder-message">
+							{ lockUrlControlsMessage }
+						</span>
+					) : (
+						isLargeContainer && content
+					) }
```
```suggestion
					{ lockUrlControls && isLargeContainer && (
						<span className="block-bindings-media-placeholder-message">
							{ lockUrlControlsMessage }
						</span>
					) }
					{ ! lockUrlControls && isLargeContainer && content }
```

Not only for `content` but also for `lockUrlControlsMessage`, I think it's better not to render the message itself if the container is not large.

Otherwise, when a small placeholder is selected, the size of the placeholder will change due to the influence of the invisible message:

https://github.com/user-attachments/assets/2aea26b9-ca4b-4949-9502-a0ae12b4f2d4

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -108,11 +108,39 @@ export function ImageEdit( {
 	const [ contentResizeListener, { width: containerWidth } ] =
 		useResizeObserver();
 
+	const [ placeholderResizeListener, { width: placeholderWidth } ] =
+		useResizeObserver();
+
+	// Parse width of the image block attribute.
+	const parseWidth = ( widthValue ) => {
+		const parsed = parseInt( widthValue, 10 );
+		return isNaN( parsed ) ? null : parsed;
+	};
+
+	const parsedWidth = parseWidth( width );
+	const parsedPlaceholderWidth = placeholderWidth || null;
+
+	// Get the effective width of the image placeholder use the minimum of placeholder width
+	// or image attribute width if both are defined. If only one is defined, use that one.
+
+	const getEffectiveWidth = () => {
+		if ( parsedPlaceholderWidth !== null && parsedWidth !== null ) {
+			return Math.min( parsedPlaceholderWidth, parsedWidth );
+		}
+		return parsedPlaceholderWidth !== null
+			? parsedPlaceholderWidth
+			: parsedWidth;
+	};
+
+	const effectiveWidth = getEffectiveWidth();
+	const isLargeContainer = effectiveWidth !== null && effectiveWidth > 160;
```
Rather than adding a new `useResizeObserver` hook, can't we reuse the one that already exists (`containerWidth`)? Also, the width/height obtained by the `useResizeObserver()` hook should only be [a number type or null type](https://github.com/WordPress/gutenberg/blob/18164f0fcd09a282157f02bba7fd737012cc44fd/packages/compose/src/hooks/use-resize-observer/index.tsx#L16-L19). Taking this into consideration, I expect the logic here can be simplified further.

vipul0425 replied:
Thanks, @t-hamano, for the review. I tried using the existing containerWidth, but since it was applied outside the `<figure>`, it always returned the content width. When using an image inside the grid, it didn't provide the exact width of the image block.
I'll update the effective width logic to account for these return types.

t-hamano replied:
Ah, I misunderstood that the existing hook was for getting the width of the block itself. I think adding a new hook makes sense 👍

t-hamano reviewed (commented):
Thanks for the update!

I'll look into the E2E test failures later, but I've left some feedback for further improvement.

vipul0425 committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feat/issue-64288

vipul0425 committed:
fix: Adressed the PR feedbacks.

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -108,11 +108,35 @@ export function ImageEdit( {
+	const [ placeholderResizeListener, { width: placeholderWidth } ] =
+		useResizeObserver();
+
+	// Parse width of the image block attribute.
+	const parseWidth = ( widthValue ) => {
+		const parsed = parseInt( widthValue, 10 );
+		return isNaN( parsed ) ? null : parsed;
+	};
+
+	const parsedWidth = parseWidth( width );
+
+	// Get the effective width of the image placeholder use the minimum of placeholder width
+	// or image attribute width if both are defined. If only one is defined, use that one.
+	const getEffectiveWidth = () => {
+		if ( placeholderWidth !== null && parsedWidth !== null ) {
+			return Math.min( placeholderWidth, parsedWidth );
+		}
+		return placeholderWidth !== null ? placeholderWidth : parsedWidth;
+	};
+
+	const effectiveWidth = getEffectiveWidth();
+	const isLargeContainer = effectiveWidth !== null && effectiveWidth > 160;
```
```suggestion
	const isLargeContainer = placeholderWidth > 160;
```

Do we need to take into account the width as an attribute? I might be missing something, but can't we simply change it like this?

vipul0425 replied:
Thanks @t-hamano I was actually using the observer outside the placeholder component that's why it was required earlier but if I move inside it then I don't think we will need that. So I will update the PR accordingly.

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -330,40 +354,48 @@ export function ImageEdit( {
+						...shadowProps.style,
+					} }
+				>
+					{ lockUrlControls && isLargeContainer && (
+						<span className="block-bindings-media-placeholder-message">
+							{ lockUrlControlsMessage }
+						</span>
```
```suggestion
					{ lockUrlControls &&
						isLargeContainer &&
						lockUrlControlsMessage }
```

Since this class name is no longer in use, the `span` tag itself is no longer necessary.

t-hamano reviewed (commented):
Thanks for the update! I left some additional feedback and I think it's pretty close to being finished 👍

I'll investigate how to solve the E2E testing issue in more detail this weekend.

vipul0425 committed:
fix: Removes the effective width logic and moves observer inside of placeholder.

jasmussen commented:
@t-hamano apologies for the direct ping, do you still have bandwidth to look at the e2e tests? Otherwise I might be able to find someone else that can help, let me know!

t-hamano commented:
@jasmussen Sorry for the late reply.

From what I can see, the E2E test failure is definitely as described in [this comment](https://github.com/WordPress/gutenberg/pull/64320#issuecomment-2323937969).

That is, after an Image block is inserted, the Upload button should be focused, as shown below:

![image](https://github.com/user-attachments/assets/d581880b-0237-4e38-9bc2-23a82b7e361b)

However, in this PR, to prevent flashing the white placeholder in narrow columns like the one below, the placeholder is initially rendered with the illustration and no focusable elements:

https://github.com/user-attachments/assets/3cb45d00-95de-4f89-a397-6ee6da7ae1be


So after the Image block is inserted, the Image block itself gets focus, not the Upload button:

![image](https://github.com/user-attachments/assets/d3cff211-581d-47a2-819e-3e94bed32bb1)

We could modify the E2E tests themselves to fit this PR, but that might not be ideal. Because it would mean making an exception to the specification that is common to all blocks: "focus the first focusable element when a block is inserted."

As of now, I have not found an approach to solve this problem, so if anyone knows a good approach on what changes to make to this PR to achieve the following, I would appreciate your help 🙇‍♂️

- In wide columns, the upload button gets focus when an image block is inserted.
- In narrow columns, the image block placeholder always renders with an illustration and does not cause a flash.

jasmussen commented:
Thanks so much for the clarification. 

I wonder: can we extract the little trick that changes the focusing to avoid the brief flash? It's one we can live without, at least initially. Fixing that flash may not be worth extra code complexity, and there might be a different approach to placeholders worth looking at separately.

jasmussen commented:
I would hate to lose the fact that the placeholder functions in narrow contexts, it's entirely broken at the moment. What I'm suggesting is we rewind back to maybe [this state of the PR](https://github.com/WordPress/gutenberg/pull/64320#issuecomment-2288133228) (if that helps), where everything seemed to work and the only problem was the brief flash of the full placeholder when inserting from the slash command in a narrow context. The reason being: that insertion is an edge-case, not too common. But the image functioning in narrow contexts is arguably common and would be fantastic to fix.

vipul0425 committed:
fix: The e2e test failures

vipul0425 commented:
Thank you so much @t-hamano for your suggestions and for helping me move this forward 🙏. I’ve made the necessary changes,  I’ll also try to address fixing the flashing issue alongside the E2E tests. If successful, I’ll create a follow-up PR for the fix.

t-hamano commented on packages/block-library/src/image/edit.js:
```diff
@@ -108,11 +108,17 @@ export function ImageEdit( {
+	const [ placeholderResizeListener, { width: placeholderWidth } ] =
+		useResizeObserver();
+
+	const isSmallContainer = placeholderWidth ? placeholderWidth < 160 : false;
```
```suggestion
	const isSmallContainer = placeholderWidth && placeholderWidth < 160;
```

Nit: I think this can be simplified.

t-hamano reviewed (approved):
Thanks for the update! I think everything is working as expected.

One last thing I noticed is that the Enter button for the "Current media URL" field is misaligned:

![image](https://github.com/user-attachments/assets/c5d3a9be-ef32-4994-b42d-b19390438254)

However, I believe this is an issue with the `MediaReplaceFlow` or `LinkControl` component itself, and may be addressed in #65158.

vipul0425 committed:
Update packages/block-library/src/image/edit.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

t-hamano commented:
It appears to have passed all E2E tests 👍
